cou/apps/auxiliary.py

```diff
@@ -0,0 +1,180 @@
+            os_origin_parsed = self.os_origin.rsplit("-", maxsplit=1)[-1]
+            match os_origin_parsed:
+                case "distro":
+                    # find the OpenStack release based on ubuntu series
+                    os_origin_parsed = LTS_SERIES[self.series]
+                case "":
+                    # if it's empty we consider the previous release from the target.
+                    # Ex: rabbitmq-server has empty "source" and receive target "victoria".
+                    # In that case it will be considered as ussuri and with the upgrade,
+                    # "source" config will be changed to "cloud:focal-victoria".
+                    os_origin_parsed = target.previous_release
+        return OpenStackRelease(os_origin_parsed) if os_origin_parsed else None
+
+    @property
+    def expected_current_channel(self) -> str:
+        """Return the expected current channel based on the current OpenStack release.
+
+        Note that this is not necessarily equal to the "channel" property since it is
+        determined based on the workload version.
+
+        :return: The expected current channel for the application.
+        :rtype: Optional[str]
+        """
+        return f"{self.openstack_to_track(self.current_os_release)}/stable"
+
+    def target_channel(self, target: OpenStackRelease) -> str:
+        """Return the channel based on the target passed.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: The next channel for the application. E.g: victoria/stable
+        :rtype: str
+        """
+        return f"{self.openstack_to_track(target)}/stable"
+
+    def _get_refresh_charm_plan(
+        self, target: OpenStackRelease, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get plan for refreshing the current channel.
+
+        This function also identifies if charm comes from charmstore and in that case,
+        makes the migration.
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan for refreshing the charm.
+        :rtype: Optional[UpgradeStep]
+        """
+        switch = None
+        channel = self.channel
+
+        description = f"Refresh '{self.name}' to the latest revision of " f"'{self.channel}'"
+
+        if self.charm_origin == "cs":
+            description = f"Migration of '{self.name}' from charmstore to charmhub"
+            switch = f"ch:{self.charm}"
+            channel = self.expected_current_channel
+        elif self.channel != self.expected_current_channel:
+            logger.warning(
+                "'%s' has the channel set to: %s that is different from the expected channel: %s",
+                self.name,
+                self.channel,
+                self.expected_current_channel,
+            )
```

Comment:
IIRC our discussion, we will not refresh the charm if its current channel doesn't match the expectation. So I think we should `return None` here

Reply:
Done

Reply:
This could lose our `idempotency`. Eg;  running the cou until change the current channel, abort and run it again. In that case we would start throwing exceptions all the time.

Reply:
you could handle that case by checking if the current channel equals the target one (in which case there'd be nothing to do), but still throw an exception in other situations

Reply:
I guess I found a way to have a unified `_get_refresh_charm_plan` for both `OpenStackApplication` and `OpenStackAuxiliaryApplication`.

For that I've changed the code to have a property named `channel_codename` that tries to get the OpenStack release by the channel. On auxiliary charms a channel can have multiple Openstack releases and in that case getting the maximum version seems to work.